servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/OptionalSslNegotiator.java

```diff
@@ -0,0 +1,125 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.netty;
+
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.http.api.HttpExecutionContext;
+import io.servicetalk.http.api.HttpServerContext;
+import io.servicetalk.http.api.StreamingHttpService;
+import io.servicetalk.tcp.netty.internal.ReadOnlyTcpServerConfig;
+import io.servicetalk.tcp.netty.internal.TcpServerBinder;
+import io.servicetalk.tcp.netty.internal.TcpServerChannelInitializer;
+import io.servicetalk.transport.api.ConnectionObserver;
+import io.servicetalk.transport.api.EarlyConnectionAcceptor;
+import io.servicetalk.transport.api.LateConnectionAcceptor;
+import io.servicetalk.transport.netty.internal.InfluencerConnectionAcceptor;
+import io.servicetalk.transport.netty.internal.NettyConnectionContext;
+
+import io.netty.channel.Channel;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.SocketAddress;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import javax.annotation.Nullable;
+
+final class OptionalSslNegotiator {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(OptionalSslNegotiator.class);
+
+    private OptionalSslNegotiator() {
+        // No instances
+    }
+
+    static Single<HttpServerContext> bind(final HttpExecutionContext executionContext,
+                                          final HttpServerConfig config,
+                                          final SocketAddress listenAddress,
+                                          @Nullable final InfluencerConnectionAcceptor connectionAcceptor,
+                                          final StreamingHttpService service,
+                                          final boolean drainRequestPayloadBody,
+                                          @Nullable final EarlyConnectionAcceptor earlyConnectionAcceptor,
+                                          @Nullable final LateConnectionAcceptor lateConnectionAcceptor) {
+        final ReadOnlyHttpServerConfig roConfig = config.asReadOnly();
+        final ReadOnlyTcpServerConfig roTcpConfig = roConfig.tcpConfig();
+
+        final BiFunction<Channel, ConnectionObserver, Single<NettyConnectionContext>> channelInit =
+                (channel, connectionObserver) ->
+                        new OptionalSslChannelSingle(channel, OptionalSslChannelSingle.NoopChannelInitializer.INSTANCE)
+                                .flatMap(isTls -> {
+                                    if (isTls) {
+                                        if (roTcpConfig.isAlpnConfigured()) {
+                                            return DeferredServerChannelBinder.alpnInitChannel(listenAddress, channel,
```

Comment:
Sad that for the same series of conditions the return result is different compare to `DefaultHttpServerBuilder.doBind` . Please add a cross comment that these 2 places must always be consistent.